source/validate_bitwise.cpp

```diff
@@ -0,0 +1,248 @@
+// Copyright (c) 2017 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Validates correctness of bitwise instructions.
+
+#include "validate.h"
+
+#include "diagnostic.h"
+#include "opcode.h"
+#include "val/instruction.h"
+#include "val/validation_state.h"
+
+namespace libspirv {
+
+namespace {
+
+// Returns operand word for given instruction and operand index.
+// The operand is expected to only have one word.
+inline uint32_t GetOperandWord(const spv_parsed_instruction_t* inst,
```

Comment:
I've seen this a few times.

A future CL could refactor this into a common spot in source/instruction.cpp.

However, the better change is a refactor of the internal representation of SPIR-V instructions, commonizing functionality between the optimizer and the validator.

So I'm happy to wait.  This is fine.

Reply:
But this is not Instruction, this is spv_parsed_instruction_t. There seems to be no good place for it.